src/compile/mark/text.ts

```diff
@@ -30,42 +35,69 @@ export namespace text {
       ['angle', 'align', 'baseline', 'dx', 'dy', 'font', 'fontWeight',
         'fontStyle', 'radius', 'theta', 'text']);
 
+    const config = model.config();
     const fieldDef = model.fieldDef(TEXT);
 
+    p.x = x(model.encoding().x, model.scaleName(X), config, fieldDef);
+
+    p.y = y(model.encoding().y, model.scaleName(Y), config, fieldDef);
+
+    p.fontSize = size(model.encoding().size, model.scaleName(SIZE), model.scale(SIZE), config);
+
+    p.text = text(model.encoding().text, model.scaleName(TEXT), config);
+
+    // FIXME: need this?
+    applyColorAndOpacity(p, model);
+    return p;
+  }
+
+  function x(fieldDefX: FieldDef, scaleName: string, config: Config, fieldDefText:FieldDef): VgValueRef {
     // x
-    if (model.has(X)) {
-      p.x = {
-        scale: model.scaleName(X),
-        field: model.field(X, { binSuffix: '_mid' })
-      };
+    if (fieldDefX) {
+      if (fieldDefX.field) {
+        return {
+          scale: scaleName,
+          field: field(X, { binSuffix: '_mid' })
+        };
+      }
     } else { // TODO: support x.value, x.datum
-      if (model.has(TEXT) && model.fieldDef(TEXT).type === QUANTITATIVE) {
-        p.x = { field: { group: 'width' }, offset: -5 };
+      if (fieldDefText && fieldDefText.type === QUANTITATIVE) {
+        return { field: { group: 'width' }, offset: -5 };
       } else {
-        p.x = { value: model.config().scale.textBandWidth / 2 };
+        return { value: config.scale.textBandWidth / 2 };
       }
     }
+  }
 
+  function y(fieldDefY: FieldDef, scaleName: string, config: Config, fieldDefText:FieldDef): VgValueRef {
     // y
-    if (model.has(Y)) {
-      p.y = {
-        scale: model.scaleName(Y),
-        field: model.field(Y, { binSuffix: '_mid' })
-      };
+    if (fieldDefY) {
+      if (fieldDefY.field) {
+        return {
+          scale: scaleName,
+          field: field(Y, { binSuffix: '_mid' })
+        };
+      }
     } else {
-      p.y = { value: model.config().scale.bandSize / 2 };
+      return { value: config.scale.bandSize / 2 };
     }
+  }
 
+  function size(fieldDefSize: FieldDef, scaleName: string, config: Config, fieldDefText:FieldDef): VgValueRef {
     // size
-    if (model.has(SIZE)) {
-      p.fontSize = {
-        scale: model.scaleName(SIZE),
-        field: model.field(SIZE)
-      };
+    if (fieldDefSize) {
+      if (fieldDefSize.field) {
+        return {
+          scale: scaleName,
+          field: field(SIZE)
+        };
+      }
     } else {
-      p.fontSize = { value: sizeValue(model) };
+      return { value: sizeValue(fieldDefSize, config) };
     }
+  }
 
+  function opacity() {
     if (model.config().mark.applyColorToBackground && !model.has(X) && !model.has(Y)) {
```

Comment:
Add comment that 

// when color is apply to background, only set opacity here.